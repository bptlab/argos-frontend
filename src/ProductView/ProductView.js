import React, { Component } from 'react';
import './ProductView.css';
import Header from './Header/Header.js';
import DetailArea from './DetailArea/DetailArea.js';
import FilterBar from './Filterbar/FilterBar.js';
import TabBar from './Tabbar/TabBar.js';
import EventTable from './EventTable/EventTable.js';

class ProductView extends Component {
    constructor(props) {
        super(props);
        const prodId = parseInt(this.props.params.productID, 10);
        this.state = {
            filter: [{id: 'filter-0', value: ''}],
            lastFilterId: 0,
            product: this.props.dataSource.receiveProduct(prodId),
            eventTypes: this.props.dataSource.receiveEventTypesOf(prodId),
            eventTable: {
                header: [],
                events: []
            }
        };
        //Function binding
        this.onChangeFilterInput = this.onChangeFilterInput.bind(this);
        this.loadEventsFor = this.loadEventsFor.bind(this);
    }

    onChangeFilterInput(currentFilterId, value) {
        if(currentFilterId === this.state.lastFilterId) {
            const newFilterId = this.state.lastFilterId + 1;
            const newFilter = {id: `filter-${newFilterId}`, value: ''};
            this.setState({ 
                filter: this.state.filter.concat([newFilter]), 
                lastFilterId: newFilterId 
            });
        } else {
            const updatedFilters = this.state.filter;
            updatedFilters[currentFilterId].value = value;
            this.setState({filter: updatedFilters});
        }
    }

    loadEventsFor(eventType) {
        const events = this.props.dataSource.receiveEventsOf(this.state.product.id, eventType.id);
        const eventTable = { 
            header: eventType.attributes, 
            events: events 
        };
        this.setState({ eventTable: eventTable });
    }

    render() {
        return (
            <div>
                <Header product={this.state.product}/>
                <DetailArea product={this.state.product}/>
                <FilterBar
                    onChangeFilterInput={this.onChangeFilterInput}
                    filter={this.state.filter}/>
                <TabBar 
                    eventTypes={this.state.eventTypes} 
                    loadEventsFor={this.loadEventsFor} 
                    product={this.state.product}/>
                <EventTable 
                    eventTable={this.state.eventTable} 
                    filter={this.state.filter}/>
            </div>
        );
    }
} export default ProductView;
