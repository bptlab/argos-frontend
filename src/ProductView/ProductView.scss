@import "DetailArea/DetailArea";
@import "EventTable/EventTable";
@import "FilterBar/FilterBar";
@import "Header/Header";
@import "TabBar/TabBar";