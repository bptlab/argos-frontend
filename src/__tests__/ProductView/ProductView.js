import React from 'react';
import ProductView from '../../ProductView/ProductView.js';
import renderer from 'react-test-renderer';
import TestProduct from './../testData/product.js'
import TestEventTypes from '../testData/frontend_eventTypes.js'
import TestEvents from '../testData/frontend_events.js'

let instance, notificationService, component;

beforeEach(() => {
    const fetchProductMockCallback = jest.fn();
    const fetchEventTypesOfMockCallback = jest.fn();
    const fetchEventsOfMockCallback = (prodId, eventTypeId, succCallback) => {
        succCallback(TestEvents.EVENTS);
    };
    notificationService = {
        subscribe: jest.fn(),
        unsubscribe: jest.fn()
    };
    component = renderer.create(
        <ProductView
            ref={(child) => {instance = child}}
            dataSource={{
                fetchProduct: fetchProductMockCallback,
                fetchEventTypesOf: fetchEventTypesOfMockCallback,
                fetchEventsOf: fetchEventsOfMockCallback,
                notificationService: notificationService
            }}
            params={{productId: 0}}/>
    );
    instance.handleProductData(TestProduct.PRODUCT);
    instance.handleEventTypeData(TestEventTypes.EVENTTYPES);
});

test("Rendering of ProductView", () => {
    const fetchProductMockCallback = jest.fn();
    const fetchEventTypesOfMockCallback = jest.fn();
    const fetchEventsOfMockCallback = (prodId, eventTypeId, succCallback) => {
        succCallback(TestEvents.EVENTS);
    };
    notificationService = {
        subscribe: jest.fn(),
        unsubscribe: jest.fn()
    };
    component = renderer.create(
        <ProductView
            ref={(child) => {instance = child}}
            dataSource={{
                fetchProduct: fetchProductMockCallback,
                fetchEventTypesOf: fetchEventTypesOfMockCallback,
                fetchEventsOf: fetchEventsOfMockCallback,
                notificationService: notificationService
            }}
            params={{productId: 0}}/>
    );
    expect(fetchProductMockCallback).toBeCalled();
    instance.handleProductData(TestProduct.PRODUCT);
    expect(fetchEventTypesOfMockCallback).toBeCalled();
    instance.handleEventTypeData(TestEventTypes.EVENTTYPES);
    
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    //Two Tabs for two eventTypes and Event+Product on ProductView
    expect(notificationService.subscribe).toHaveBeenCalledTimes(3);
});

test('Update Event data', () => {
    //Use same Eventtype again to check other if branch
    instance.handleEventData(TestEventTypes.EVENTTYPES[0], TestEvents.EVENTS);
});

test('Successful filtering', () => {
    instance.onInputChange(0, 'Product');
    expect(instance.state.filter[0].value).toBe("Product");
    expect(instance.state.filter[0].column).toBeNull();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Changing active Tab', () => {
    instance.setActiveEventType(TestEventTypes.EVENTTYPES[0]);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Change filter', () => {
    instance.onInputChange(0, 'columnName:Product');
    expect(instance.state.filter[0].value).toBe("Product");
    expect(instance.state.filter[0].column).toBe("columnName");
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Add filter', () => {
    expect(instance.state.filter.length).toEqual(1);
    instance.onInputChange(0, 'Test');
    expect(instance.state.filter.length).toEqual(2);
});

test('Remove last filter', () => {
    expect(instance.state.filter.length).toEqual(1);
    instance.onInputChange(0, 'Test');
    expect(instance.state.filter.length).toEqual(2);

    instance.onInputChange(0, '');
    expect(instance.state.filter.length).toEqual(1);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Remove intermediate filter', () => {
    expect(instance.state.filter.length).toEqual(1);
    instance.onInputChange(0, 'Test');
    expect(instance.state.filter.length).toEqual(2);
    instance.onInputChange(1, 'Test2');
    expect(instance.state.filter.length).toEqual(3);

    instance.onInputChange(0, '');
    expect(instance.state.filter.length).toEqual(2);
    expect(instance.state.filter[0].value).toEqual('Test2');
    expect(instance.state.filter[1].value).toEqual('');
});

test('Handle error response', () => {
    instance.handleError(404);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Unmounting component', () => {
    instance.componentWillUnmount();
    expect(notificationService.unsubscribe).toHaveBeenCalledTimes(2);
});
