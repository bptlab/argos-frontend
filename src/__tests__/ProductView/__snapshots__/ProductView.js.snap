exports[`test Rendering of ProductView 1`] = `
<div>
  <div>
    <header>
      <div
        className="container">
        <div
          className="row d-flex align-items-center">
          <div
            className="hidden-xs-down col-sm-1 d-flex justify-content-center">
            <a
              href="/">
              <i
                className="home-button fa fa-home" />
            </a>
          </div>
          <div
            className="col-xs-12 col-sm-10 d-flex justify-content-around">
            <h1
              className="dashboard-title">
              Testproduct 30X 1W1-PSB
            </h1>
          </div>
          <div
            className="hidden-xs-down col-sm-1 d-flex justify-content-around">
            <div
              className="status-point d-flex align-items-center error">
              <svg
                className="status-point"
                viewBox="0 0 42 42">
                <circle
                  className="status-point-shadow"
                  cx="21"
                  cy="21"
                  r="21" />
                <circle
                  className="status-point"
                  cx="20"
                  cy="20"
                  r="20" />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      className="query-interface">
      <div>
        <button
          className="btn btn-primary"
          data-target="#query-interface-modal"
          data-toggle="modal"
          type="button">
          Create new query
        </button>
        <div
          className="query-interface modal fade"
          id="query-interface-modal"
          role="dialog"
          tabIndex="-1">
          <div
            className="modal-dialog"
            role="document">
            <div
              className="modal-content">
              <div
                className="modal-header">
                <h5
                  className="modal-title"
                  id="query-interface-label">
                  Create new query
                </h5>
                <button
                  className="close"
                  data-dismiss="modal"
                  type="button">
                  <span>
                    ×
                  </span>
                </button>
              </div>
              <div
                className="modal-body">
                <div
                  className="query-interface-header">
                  <div
                    className="form-group">
                    <input
                      className="form-control"
                      id="query-name"
                      placeholder="Name"
                      type="text" />
                  </div>
                  <div
                    className="query-interface-header-attribute form-group row">
                    <div
                      className="col-6">
                      <input
                        className="form-control"
                        id="example-text-input"
                        onChange={[Function]}
                        placeholder="Attribute name"
                        type="text"
                        value="timestamp" />
                    </div>
                    <div
                      className="col-6">
                      <select
                        className="form-control"
                        id="exampleSelect1"
                        onChange={[Function]}
                        value="Date">
                        <option>
                          String
                        </option>
                        <option>
                          Boolean
                        </option>
                        <option>
                          Integer
                        </option>
                        <option>
                          Float
                        </option>
                      </select>
                    </div>
                  </div>
                  <div
                    className="query-interface-header-attribute form-group row">
                    <div
                      className="col-6">
                      <input
                        className="form-control"
                        id="example-text-input"
                        onChange={[Function]}
                        placeholder="Attribute name"
                        type="text"
                        value="productId" />
                    </div>
                    <div
                      className="col-6">
                      <select
                        className="form-control"
                        id="exampleSelect1"
                        onChange={[Function]}
                        value="Integer">
                        <option>
                          String
                        </option>
                        <option>
                          Boolean
                        </option>
                        <option>
                          Integer
                        </option>
                        <option>
                          Float
                        </option>
                      </select>
                    </div>
                  </div>
                  <div
                    className="query-interface-header-attribute form-group row">
                    <div
                      className="col-6">
                      <input
                        className="form-control"
                        id="example-text-input"
                        onChange={[Function]}
                        placeholder="Attribute name"
                        type="text"
                        value="productFamilyId" />
                    </div>
                    <div
                      className="col-6">
                      <select
                        className="form-control"
                        id="exampleSelect1"
                        onChange={[Function]}
                        value="String">
                        <option>
                          String
                        </option>
                        <option>
                          Boolean
                        </option>
                        <option>
                          Integer
                        </option>
                        <option>
                          Float
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
                <div
                  className="form-group">
                  <label
                    className="form-control-label"
                    htmlFor="event-query">
                    Event Query
                  </label>
                  <textarea
                    className="form-control"
                    id="event-query"
                    onChange={[Function]}
                    rows="8"
                    type="text"
                    value="INSERT INTO {NAME}[timestamp: Date, productId: Integer, productFamilyId: String] SELECT timestamp, productId, ProductFamilyId FROM ErrorEvents" />
                </div>
                <p>
                  
                </p>
              </div>
              <div
                className="modal-footer">
                <button
                  className="btn btn-primary"
                  onClick={[Function]}
                  type="button">
                  Save
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="detail-area container">
      <div
        className="row detail-row">
        <div
          className="detail-area-block col-xs-12 col-sm-6">
          <div>
            <div
              className="row">
              <h2
                className="col area-title">
                Produkt Details
              </h2>
            </div>
            <div
              className="row">
              <p
                className="col-4">
                Marke:
              </p>
              <p
                className="col">
                Testbrand1
              </p>
            </div>
            <div
              className="row">
              <p
                className="col-4">
                Produktfamilie:
              </p>
              <p
                className="col">
                Testproduct 30X 1W1-PSB
              </p>
            </div>
            <div
              className="row">
              <p
                className="col-4">
                Seriennummer:
              </p>
              <p
                className="col">
                0
              </p>
            </div>
          </div>
        </div>
        <div
          className="detail-area-block col-xs-12 col-sm-6">
          <div>
            <h2
              className="area-title">
              Mögliche Fehlerursachen
            </h2>
            <ol>
              <li>
                <h4>
                  ...
                </h4>
              </li>
              <li>
                <h4>
                  ...
                </h4>
              </li>
              <li>
                <h4>
                  ...
                </h4>
              </li>
            </ol>
          </div>
        </div>
      </div>
    </div>
    <div
      className="filter-bar container">
      <div
        className="form-group row">
        <div
          className="filter input-group col-xs-12 col-sm-4 col-lg-3">
          <span
            className="input-group-addon"
            id="search-addon">
            <i
              className="fa fa-search" />
          </span>
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value={undefined} />
        </div>
      </div>
    </div>
    <div
      className="tab-bar container">
      <ul
        className="nav nav-tabs border-0"
        role="tablist">
        <li
          className="tab nav-item">
          <a
            className="nav-link"
            data-toggle="tab"
            href="#Error Events"
            onClick={[Function]}
            role="tab">
            Error Events
          </a>
        </li>
      </ul>
    </div>
    <div
      className="container event-table-container">
      <table
        className="table">
        <thead
          className="event-table-header thead">
          <tr>
            <th>
              #
            </th>
            <th>
              ErrorCode
            </th>
            <th>
              ErrorDescription
            </th>
            <th>
              PossibleFix
            </th>
            <th>
              DeviceID
            </th>
            <th>
              DeviceLocation
            </th>
            <th>
              SoftwareVersion
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th
              scope="row">
              0
            </th>
            <td>
              404
            </td>
            <td>
              Product not found.
            </td>
            <td>
              Search for it.
            </td>
            <td>
              666
            </td>
            <td>
              Not known.
            </td>
            <td>
              0.0.4.alpha
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`test Rendering of ProductView 2`] = `
<div>
  <div
    className="critical-error">
    <i
      className="fa fa-exclamation-triangle warning-sign" />
    <p>
      404
    </p>
  </div>
</div>
`;
